### PR TITLE
[KMSv2] use encDEK, keyID and annotations to generate cache key

### PR DESCRIPTION
**Pull request by aramase**
Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
It is possible for a KMSv2 plugin to return a static value as Ciphertext
and store the actual encrypted DEK in the annotations. In this case,
using the encDEK will not work. Instead, we are now using a combination
of the encDEK, keyID and annotations to generate the cache key.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by valaparthvi** [comment]
Hi @aramase, I am from the Bug Triage Team for K8s release v1.27. I see this PR is in actively worked up, so I assume it is on track for this release.

However, please note that Code Freeze begins on March 15, after which only critical and release blocker PR/issues will be accepted. Please make sure this PR goes in by March 15.

Do let us know if you need help in getting this PR reviewed, we can work something up :)

**Comment by dims** [comment]
@aramase please rebase!

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go [hunk 1]
```diff
@@ -861,6 +873,107 @@ func TestEnvelopeLogging(t *testing.T) {
 	}
 }
 
+func TestCacheNotCorrupted(t *testing.T) {
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:876]: This test fails in master as we're only using the `EncryptedDEK` as cache key.

```bash
    /home/anramase/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:926: cipher: message authentication failed
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go [hunk 2]
```diff
@@ -98,5 +98,11 @@ func (c *simpleCache) keyFunc(s []byte) string {
...
 // toString performs unholy acts to avoid allocations
 func toString(b []byte) string {
-	return *(*string)(unsafe.Pointer(&b))
+	// unsafe.SliceData relies on cap whereas we want to rely on len
+	if len(b) == 0 {
+		return ""
+	}
+	// Copied from go 1.20.1 strings.Builder.String
+	// https://github.com/golang/go/blob/202a1a57064127c3f19d96df57b9f9586145e21c/src/strings/builder.go#L48
+	return unsafe.String(unsafe.SliceData(b), len(b))
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go:107]: These changes have been mirrored from https://github.com/kubernetes/kubernetes/pull/115935

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 3]
```diff
@@ -339,3 +355,58 @@ func getRequestInfoFromContext(ctx context.Context) *genericapirequest.RequestIn
...
+// toBytes performs unholy acts to avoid allocations
+func toBytes(s string) []byte {
+	// unsafe.StringData is unspecified for the empty string, so we provide a strict interpretation
+	if len(s) == 0 {
+		return nil
+	}
+	// Copied from go 1.20.1 os.File.WriteString
+	// https://github.com/golang/go/blob/202a1a57064127c3f19d96df57b9f9586145e21c/src/os/file.go#L246
+	return unsafe.Slice(unsafe.StringData(s), len(s))
+}
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:412]: These changes have been mirrored from https://github.com/kubernetes/kubernetes/pull/115935

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 4]
```diff
@@ -339,3 +355,58 @@ func getRequestInfoFromContext(ctx context.Context) *genericapirequest.RequestIn
...
+		return b.Bytes()
+	}
+
+	// add the length of annotations to the cache key
+	b.AddUint32(uint32(len(annotations)))
+
+	// Sort the annotations by key.
+	keys := make([]string, 0, len(annotations))
+	for k := range annotations {
+		keys = append(keys, k)
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:386]: I think the existing code is fine but this makes me feel better:

```suggestion
	for k := range annotations {
		k := k
		keys = append(keys, k)
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 5]
```diff
@@ -137,8 +143,13 @@ func (t *envelopeTransformer) TransformFromStorage(ctx context.Context, data []b
...
 	}
 
+	encryptedObjectCacheKey, err := generateCacheKey(encryptedObject.EncryptedDEK, encryptedObject.KeyID, encryptedObject.Annotations)
+	if err != nil {
+		return nil, false, err
+	}
+
 	// Look up the decrypted DEK from cache first
-	transformer := t.cache.get(encryptedObject.EncryptedDEK)
+	transformer := t.cache.get(encryptedObjectCacheKey)
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:152]: The call below to `addTransformerForDecryption` is incorrect - please add a test that catches that :)

**Review by enj — CHANGES_REQUESTED** [review]
Still looking through the tests.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go (hunk 5 above)
Comment (aramase) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:152]: good catch! interesting how the current tests didn't catch it. Let me check!

Comment (aramase) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:152]: Added a new test in `TestEnvelopeCaching` and modified it to do multiple reads to test this. The new table test failed as expected to catch this

```bash
=== RUN   TestEnvelopeCaching
=== RUN   TestEnvelopeCaching/entry_in_cache_should_withstand_plugin_failure
=== RUN   TestEnvelopeCaching/cache_entry_expired_should_not_withstand_plugin_failure
=== RUN   TestEnvelopeCaching/cache_entry_expired_should_work_after_cache_refresh
    /home/anramase/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:221: expected 1 decrypt calls, got 10
--- FAIL: TestEnvelopeCaching (0.00s)
    --- PASS: TestEnvelopeCaching/entry_in_cache_should_withstand_plugin_failure (0.00s)
    --- PASS: TestEnvelopeCaching/cache_entry_expired_should_not_withstand_plugin_failure (0.00s)
    --- FAIL: TestEnvelopeCaching/cache_entry_expired_should_work_after_cache_refresh (0.00s)
FAIL
FAIL	k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2	0.005s
```

**Pull request merged**